cosmwasm/register-its-token.js

```diff
@@ -0,0 +1,128 @@
+            if(chainName === this.token().originChain.origin_chain) continue;
+            const current = this.get();
+            if (!current.registered) {
+                if (current.track) try {
+                    const provider = getDefaultProvider(this.rpc());
+                    const token = new Contract(current.tokenAddress, IInterchainToken.abi, provider);
+                    current.supply = await token.totalSupply();
+                } catch (e) {
+                    printError('Failed to query token supply for', current.tokenAddress);
+                }
+                console.log(`Chain Progress: ${this.chainIndex + 1}/${this.chainNames.length} | Token Progress: ${this.tokenIndex + 1}/${this.tokenIds.length}`)
+                return true;
+            }
+        }
+    }
+}
+
+async function registerToken(tokenIterator) {
+    const config = loadConfig(env);
+
+    initContractConfig(config, {contractName: "InterchainTokenService"});
+
+    const wallet = await prepareWallet({mnemonic});
+    const client = await prepareClient(config, wallet);
+    const supply = tokenIterator.get().supply;
+    const supplyParam = supply ? {"tracked": String(supply)} : "untracked";
+    const msg = { "register_p2p_token_instance": {
+        "chain": tokenIterator.chainName(),
+        "token_id": tokenIterator.tokenId().slice(2),
+        "origin_chain": tokenIterator.token().originChain,
+        "decimals": tokenIterator.get().decimals,
+        "supply": supplyParam,
+    } };
+    const registeredChains = {};
+    const registered = await client.queryContractSmart(info.axelar.contracts.InterchainTokenService.address, {"token_instance": {chain: tokenIterator.chainName(), "token_id": tokenIterator.tokenId().slice(2)}});
+    if (registered) return;
+    const [account] = await wallet.getAccounts(info.axelar.contracts.InterchainTokenService.address, {});
+
+    const ensureChainRegistered = async (chain) => {
+        if(registeredChains[chain]) return;
+        registeredChains[chain] = true;
+        if(await client.queryContractSmart(info.axelar.contracts.InterchainTokenService.address, {"its_chain": {"chain": chain}})) return;
+        const msg = {"register_chains":{"chains":[{"chain":chain,"its_edge_contract":"source-its-contract","truncation":{"max_uint_bits":256,"max_decimals_when_truncating":255}}]}};
+        await client.execute(account.address, info.axelar.contracts.InterchainTokenService.address, msg, 'auto');
+    }
+    await ensureChainRegistered(tokenIterator.chainName());
+    await ensureChainRegistered(tokenIterator.token().originChain);
+    
+    await client.execute(account.address, info.axelar.contracts.InterchainTokenService.address, msg, 'auto');
+    // If registration is successfull skip this token in the future without needing to query.
+    tokenIterator.get().registered = true;
+
+    saveConfig(config, env);
+}
+
+if (require.main === module) {
+    (async () => {
```

Comment:
the script should be able to register one or more or all provided tokenIds for a provided chain based on the usual cli input

Reply:
done